### ISSUE TITLE
Can't make a connection to some hosts; no initial data Error

### ISSUE THREAD

**Issue by hardwarefresser** [OWASP/O-Saft#44]
I also posted this on the Mailing list. On some hosts (for example httpsnow.org) the scan terminates with an error:

**WARNING: Can't make a connection to httpsnow.org:443; no initial data
do_ssl_open() WARNING 'Net::SSLeay::CTX_set_ssl_version()' not  available, using system default
do_ssl_open() failed calling Net::SSLeay::connect()  failed handshake:
do_ssl_open() WARNING 'Net::SSLeay::CTX_set_ssl_version()' not available, using system default
do_ssl_open() failed calling Net::SSLeay::connect()  failed handshake:
**WARNING: Can't make a connection to httpsnow.org:443; target ignored

When I run the scan on other sites (google.de) it works fine without any error. O-Saft seems to terminate the whole scan as soon as the connection fails.
When I uncomment the "goto CLOSE_SSL;" (line 5968) the scan finishes and the cipher check works fine, but checks like +sts or +certificate don't work and another error appears: Use of uninitialized value in string eq at o-saft.pl line 4054

My o-saft call looks like this: perl o-saft.pl httpsnow.org +hsts +certificate --no-dtlsv1 --v

**Comment by EnDe** [comment]
Reason for the WARNING is that your underlaying libssl does not support SSLv23_method(). Can you please post output from:  o-saft.pl +version
(need at least the information about openssl)

You don't need to change the code, adding --no-tlsv13 --no-tlsv12 should also help in your case.

**Comment by hardwarefresser** [comment]
Thanks for the reply!

o-saft.pl +version output:

    o-saft.pl 15.05.16
    openssl version (ext executable) OpenSSL 1.0.1k 8 Jan 2015
    Net::SSLeay::SSLeay_version()    OpenSSL 1.0.1k 8 Jan 2015
    Net::SSLeay::
    ::OPENSSL_VERSION_NUMBER()    0x268439743
    ::SSLeay()                    0x268439743
    = Required (and used) Modules =
    IO::Socket::INET     1.35
    IO::Socket::SSL      2.020
    Net::DNS             0.81
    Net::SSLinfo         15.05.16
    Net::SSLhello        15.01.26
    Net::SSLeay          1.72

Seems like my openssl supports TLSv1.2, but with the TLS_client_method() instead of the SSLv23_method().
Adding --no-tlsv13 is ok since my openssl doesn't support TLS1.3. But If I add --no-tlsv12 I can't check TLS1.2 anymore, this world be bad.

**Comment by EnDe** [comment]
ok, I see: Net::SSLeay does no longer provide some functions by default.
Please try to adapt SSLeay.xs as follows:

-----<<<<-----
#ifndef OPENSSL_NO_SSL2
const SSL_METHOD *
SSLv2_method()
#endif

#ifndef OPENSSL_NO_SSL3
const SSL_METHOD *
SSLv3_method()

const SSL_METHOD *
SSLv23_method()
#endif
----->>>>-----

This means that you remove the
  #if OPENSSL_VERSION_NUMBER < 0x10002000L
condition and add  SSLv23_method()
You also don't need to remove the
  #ifndef OPENSSL_NO_SSL2
  #ifndef OPENSSL_NO_SSL3
conditions (as long as you don't use these defines).

Then run   make&&make install.

Note: this makes the API to this functions available, so they must be in libssl.* otherwise you most likely get a Segmentation fault when using them.

Please let me know if it helps

**Comment by hardwarefresser** [comment]
Okay. My SSLeay.xs now looks like this. I have not done any changes at the SSL CONTEXT functions:
```
#define REM1 "============= SSL CONTEXT functions =============="

SSL_CTX *
SSL_CTX_new()
     CODE:
     RETVAL = SSL_CTX_new (SSLv23_method());
     OUTPUT:
     RETVAL

#ifndef OPENSSL_NO_SSL2

SSL_CTX *
SSL_CTX_v2_new()
     CODE:
     RETVAL = SSL_CTX_new (SSLv2_method());
     OUTPUT:
     RETVAL

#endif

#ifndef OPENSSL_NO_SSL3

SSL_CTX *
SSL_CTX_v3_new()
     CODE:
     RETVAL = SSL_CTX_new (SSLv3_method());
     OUTPUT:
     RETVAL

#endif

SSL_CTX *
SSL_CTX_v23_new()
     CODE:
     RETVAL = SSL_CTX_new (SSLv23_method());
     OUTPUT:
     RETVAL
```

At this sections I did the changes you suggested. I deleted the #if OPENSSL_VERSION_NUMBER < 0x10002000L and added the SSLv2_method()
````
#ifndef OPENSSL_NO_SSL2
#if OPENSSL_VERSION_NUMBER < 0x10000000L

const SSL_METHOD *
SSLv2_method()

#endif
#endif

#ifndef OPENSSL_NO_SSL3

const SSL_METHOD *
SSLv3_method()

const SSL_METHOD *
SSLv23_method()

#endif

const SSL_METHOD *
TLSv1_method()

````

To compile Net::SSLeay, I used "OPENSSL_PREFIX=/root/osaft/ssl perl Makefile.PL" followed by make&&make test&&make install
I set the OPENSSL_PREFIX because my global openssl does not support SSLv2. So I compiled a dedicated openssl for O-Saft (when I check the linked libssl of my Net::SSLeay via ldd everything looks good:)
```
linux-vdso.so.1 (0x00007ffe011e8000)
libssl.so.1.0.0 => /root/osaft/ssl/lib/libssl.so.1.0.0 (0x00007f01c541c000)
libcrypto.so.1.0.0 => /root/osaft/ssl/lib/libcrypto.so.1.0.0 (0x00007f01c5031000)
libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f01c4e16000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f01c4a6d000)
libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f01c4869000)
/lib64/ld-linux-x86-64.so.2 (0x00007f01c58fa000)
```

Unfortunately, as you promised, I get a Segmentation fault when running O-Saft. At the moment, I don't know why because according to the openssl docs my openssl v1.0.1k should support the SSLv23_method() (https://www.openssl.org/docs/man1.0.1/ssl/SSL_CTX_new.html)

**Comment by EnDe** [comment]
fFxed a bug in Net/SSLinfo.pm which also may cause the Segmentation fault.
Please see: https://github.com/OWASP/O-Saft/archive/master.zip

**Comment by EnDe** [comment]
Just released VERSION 15.11.15 which fixes some minor bugs acording described problem above. This version also tries to detect available SSL/TLS versions better and without bothering with perl warnings.

Can you please confirm that this version works with a properly installed Net::SSLeay (as described above https://github.com/OWASP/O-Saft/issues/44#issuecomment-151395387

Thanks

**Comment by hardwarefresser** [comment]
Please excuse my late repy! Will check it asap
